### PR TITLE
refactor: Public key resolver returning key type

### LINKED ISSUE #1446: Public key resolver returning key type
In `signature` and `verifiable` packages, we use a public key resolver abstraction that takes public key URI (e.g. DID) and resolves it into a public key ([]byte). 
In some cases, having public key bytes is not enough. For example, for KMS/Crypto based on https://github.com/google/tink, we need to convert the public key to key handle via exporting, the opposite operation is needed as well.

### PR DESCRIPTION
**Pull request by kdimak**
closes #1446

Signed-off-by: Dmitriy Kinoshenko <dkinoshenko@gmail.com>

We will use the public key type in order to get KMS handle. A prototype function declaration from @Baha-sk :

```
// CreatePublicKeyHandle will create and return a key handle for pubKey of type kt
// it returns an error if it failed creating the key handle
// Note: The key handle created is not stored in the KMS
func (l *LocalKMS) CreatePublicKeyHandle(pubKey []byte, kt string)
```

This PR is a one of the preparation steps to use KMS/Crypto based signature verifiers in `verifiable`.

### CODE REVIEW

File: pkg/didcomm/protocol/didexchange/states.go [hunk 1]
```diff
@@ -591,7 +591,7 @@ func verifySignature(connSignature *ConnectionSignature, recipientKeys string) (
 	// TODO: Replace with signed attachments issue-626
 	signatureSuite := ed25519signature2018.New()
 
-	err = signatureSuite.Verify(pubKey, sigData, signature)
+	err = signatureSuite.Verify("ed25519", pubKey, sigData, signature)
```
Comment (sandrask) [inline pkg/didcomm/protocol/didexchange/states.go]: I would prefer to have PublicKey struct that contains Type and Value. Both type and value are properties of public key. Later on if we need additional properties it will be easy to add. @Baha-sk

Comment (Baha-sk) [reply pkg/didcomm/protocol/didexchange/states.go]: I'm also working on constraining the key types as constants in the KMS, we can reference these types instead of open string value (once my PR is merged you can take advantage of it).

Comment (kdimak) [reply pkg/didcomm/protocol/didexchange/states.go]: Agree, @sandrask .  I will refactor that.

File: pkg/doc/verifiable/support_test.go [hunk 2]
```diff
@@ -202,3 +202,12 @@ func (s rs256TestSigner) Sign(data []byte) ([]byte, error) {
 
 	return rsa.SignPKCS1v15(rand.Reader, s.privKey, crypto.SHA256, hashed)
 }
+
+const (
+	rsaPublicKeyType     = "RSA"
+	ed25519PublicKeyType = "ed25519"
```
Comment (Baha-sk) [inline pkg/doc/verifiable/support_test.go]: probably expose these key types or better yet, use KeyType defined in pkg/kms/api.go as it should be the official key type in Aries-Framework-Go

and replace all references of "ed25519" in this patch.. there's at least 12, but in different packages than `verifiable`

File: pkg/kms/legacykms/kms_test.go [hunk 3]
```diff
@@ -10,6 +10,7 @@ import (
 	"crypto/rand"
 	"encoding/json"
 	"fmt"
+	"github.com/hyperledger/aries-framework-go/pkg/doc/signature/verifier"
```
Comment (Baha-sk) [inline pkg/kms/legacykms/kms_test.go]: move this import under line 20 below

Comment (kdimak) [reply pkg/kms/legacykms/kms_test.go]: that's why the lint check failed... rebased

File: pkg/doc/verifiable/support_test.go (hunk 2 above)
Comment (kdimak) [reply pkg/doc/verifiable/support_test.go]: Makes sense! I added `PublicKeyType` in a way similar to `kms.KeyType`.

File: pkg/doc/signature/verifier/verifier.go [hunk 4]
```diff
@@ -23,17 +23,34 @@ type signatureSuite interface {
...
+// PublicKeyType is a type of a public key
+type PublicKeyType string
+
+const (
+	// RSAPublicKeyType public key type
+	RSAPublicKeyType = PublicKeyType("RSA")
+
+	// Ed25519PublicKeyType public key type
+	Ed25519PublicKeyType = PublicKeyType(kms.Ed25519Type)
+)
```
Comment (Baha-sk) [inline pkg/doc/signature/verifier/verifier.go]: Instead of creating new key type struct here, add `RSA` key type in kms/api.go and reference its KeyTypes directly.

We should have only one keyType struct across the framework to avoid duplication.

Comment (kdimak) [reply pkg/doc/signature/verifier/verifier.go]: OK, I wil change that!

Comment (kdimak) [reply pkg/doc/signature/verifier/verifier.go]: changed

File: pkg/doc/verifiable/linked_data_proof.go [hunk 5]
```diff
@@ -38,7 +37,7 @@ type verifierSignatureSuite interface {
 	signatureSuite
 
 	// Verify will verify signature against public key
-	Verify(pubKey []byte, doc []byte, signature []byte) error
+	Verify(pubKey *verifier.PublicKey, doc []byte, signature []byte) error
```
Comment (Baha-sk) [inline pkg/doc/verifiable/linked_data_proof.go]: add a TODO to replace `verifier.PublicKey` with `keyset.Handle`.

`Verify()` implementations should call pkg/crypto's Verify() function with the instance of keyset.Handle passed here.

And the same thing for Sign() implementations, they should use pkg/crypto's default Sign() function that accepts a keyset.Handle as argument.

File: pkg/doc/verifiable/linked_data_proof.go [hunk 6]
```diff
@@ -52,25 +51,20 @@ type keyResolverAdapter struct {
 	pubKeyFetcher PublicKeyFetcher
 }
 
-func (k *keyResolverAdapter) Resolve(id string) ([]byte, error) {
+func (k *keyResolverAdapter) Resolve(id string) (*verifier.PublicKey, error) {
```
Comment (Baha-sk) [inline pkg/doc/verifiable/linked_data_proof.go]: same here... add a TODO comment that says: replace `*verifier.PublickKey` with Tink's `*keyset.Handle`

File: pkg/doc/verifiable/linked_data_proof.go (hunk 5 above)
Comment (kdimak) [reply pkg/doc/verifiable/linked_data_proof.go]: I think we wanted to don't make tight coupling with crypto/kms? Just make generic interfaces of signer / verifier without passing a private key as an argument to signer? @troyronda

Comment (Baha-sk) [reply pkg/doc/verifiable/linked_data_proof.go]: or if we don't want to limit our APIs to Tink, then make the pubKey argument to be `interface{}` type.

and default implementation will use `*keyset.Handle` Tink type

other implementations requiring []byte can use `*verifier.PublicKey`

Comment (Baha-sk) [reply pkg/doc/verifiable/linked_data_proof.go]: > I think we wanted to don't make tight coupling with crypto/kms? Just make generic interfaces of signer / verifier without passing a private key as an argument to signer? @troyronda

the concept of hiding keys entails using a key reference to execute a crypto primitive, which should be resolved by a kms implementation and primitive executed by a crypto implementation.

in order to not tie the code to any specific implementation, we should simply use `interface{}` type for keys. This way, kms/crypto Tink implementations can exchange `*keyset.Handle` with verifierSignatureSuite/signerSignatureSuite implemented with Tink key handles. Other KMS implementations will exchange their own key type with their respective signature suite implementations.

Comment (kdimak) [reply pkg/doc/verifiable/linked_data_proof.go]: I see your point. Thank you for the explanation! I created https://github.com/hyperledger/aries-framework-go/issues/1458.

File: pkg/doc/verifiable/example_presentation_test.go [hunk 7]
```diff
@@ -90,7 +97,7 @@ func ExamplePresentation_JWTClaims() {
 `
 
 	// The Holder wants to send the presentation to the Verifier in JWS.
-	vp, err := verifiable.NewPresentation([]byte(vpStrFromWallet))
+	vp, err := verifiable.NewUnverifiedPresentation([]byte(vpStrFromWallet))
```
Comment (troyronda) [inline pkg/doc/verifiable/example_presentation_test.go]: Why this change?

File: pkg/doc/verifiable/presentation.go [hunk 8]
```diff
@@ -306,6 +306,27 @@ func NewPresentation(vpData []byte, opts ...PresentationOpt) (*Presentation, err
 		return nil, err
 	}
 
+	return newPresentation(vpRaw, vpOpts)
+}
+
+// NewUnverifiedPresentation decodes Verifiable Presentation from bytes which could be marshalled JSON or
+// serialized JWT. It does not make a proof check though. Can be used for purposes of decoding of VP stored in a wallet.
+// Please use this function with caution.
+func NewUnverifiedPresentation(vpBytes []byte) (*Presentation, error) {
```
Comment (troyronda) [inline pkg/doc/verifiable/presentation.go]: We should have a TODO to re-do naming (e.g., with Parse prefix).

File: pkg/doc/verifiable/example_presentation_test.go (hunk 7 above)
Comment (kdimak) [reply pkg/doc/verifiable/example_presentation_test.go]: `vpStrFromWallet` has no embedded proof. If we pass it to `NewPresentation()` as JSON bytes, it will raise an error saying that no proof was found.
This test did not fail before because of `//nolint:lll` before `//Output:` line which made the test function non-effective. I moved `//nolint:lll` to the line before the method declaration to enable the test. There were about 10 not working example-based tests in `verifiable`.

File: pkg/doc/verifiable/presentation.go (hunk 8 above)
Comment (kdimak) [reply pkg/doc/verifiable/presentation.go]: Yeah, I will file the issue soon.
